### DISCUSSION TITLE
Adding a Custom Layer?

### DISCUSSION THREAD

**Discussion by tmdarwen** [OE4T/meta-tegra#881] [Q&A] [answered]
I'm attempting to add a custom layer to my yocto-tegra image and running into some problems.  I'm following the Yocto manual instructions [here](https://docs.yoctoproject.org/singleindex.html#creating-your-own-general-layer).

I'm able to run the `bitbake-layers create-layer` command as shown below without any problem:
```
buildmachine@buildmachine:~/yocto-tegra$ bitbake-layers create-layer meta-custom-layer
NOTE: Starting bitbake server...
Add your new layer with 'bitbake-layers add-layer meta-custom-layer'
buildmachine@buildmachine:~/yocto-tegra$ ls
build  deploy.sh  meta-custom-layer  meta-tegra  poky-dunfell
```

As instructed, I then run `bitbake-layers add-layer` as shown below but it results in an error:
```
buildmachine@buildmachine:~/yocto-tegra$ bitbake-layers add-layer meta-custom-layer
NOTE: Starting bitbake server...
Unable to find bblayers.conf
```

Are there some specific instructions for adding a custom layer to a yocto-tegra build that differ from the Yocto Manual instructions?

**Comment by madisongh** [comment]
No, there's nothing special.  The instructions in the manual may not be 100% complete, though.  Try
```
$ cd $BUILDDIR
$ bitbake-layers add-layer /path/to/meta-custom-layer
```
Most of the bitbake commands only work if your current working directory is the build directory (that's the `$BUILDDIR` reference above).

**Comment by tmdarwen** [comment]
Ah, got it, thanks, that did it.  So, I've added a simple "hello world" C++ program, but when building I'm getting the errors shown below.  

I've done this exact same thing with a Yocto build on a Raspberry Pi Zero recently without any problems.

Do you know if the yocto-tegra build includes the proper dependencies necessary for building a simple C++ program?

The BBLAYERS in my yocto-tegra project are:
```
  /home/${USER}/yocto-tegra/meta-tegra \
  /home/${USER}/yocto-tegra/poky-dunfell/meta \
  /home/${USER}/yocto-tegra/poky-dunfell/meta-poky \
  /home/${USER}/yocto-tegra/poky-dunfell/meta-yocto-bsp \
  /home/${USER}/yocto-tegra/meta-custom-layer \
```

The BBLAYERS in my yocto raspberry pi project were:
```
   /home/${USER}/yocto/meta-raspberrypi \
   /home/${USER}/yocto/meta-openembedded/meta-oe \
   /home/${USER}/yocto/meta-openembedded/meta-python \
   /home/${USER}/yocto/meta-openembedded/meta-networking \
   /home/${USER}/yocto/meta-custom-layer \
```
I wonder if I need to include meta-openembedded/meta-oe in my yocto-tegra build in order to properly build my C++ custom layer?  Any idea?

```
ERROR: core-image-sato-dev-1.0-r0 do_rootfs: Could not invoke dnf. Command '/home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/recipe-sysroot-native/usr/bin/dnf -v --rpmverbosity=info -y -c /home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/rootfs/etc/dnf/dnf.conf --setopt=reposdir=/home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/rootfs/etc/yum.repos.d --installroot=/home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/rootfs --setopt=logdir=/home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/temp --repofrompath=oe-repo,/home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/oe-rootfs-repo --nogpgcheck install dnf example gstreamer1.0-omx-tegra gstreamer1.0-plugins-nvvideo4linux2 packagegroup-core-ssh-dropbear packagegroup-core-x11-base packagegroup-core-x11-sato psplash rpm run-postinsts locale-base-en-us locale-base-en-gb' returned 1:
DNF version: 4.2.2
cachedir: /home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/rootfs/var/cache/dnf
Added oe-repo repo from /home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/oe-rootfs-repo
repo: using cache for: oe-repo
not found other for:
not found modules for:
not found deltainfo for:
not found updateinfo for:
oe-repo: using metadata from Tue 11 Jan 2022 10:37:44 PM UTC.
No module defaults found
--> Starting dependency resolution
--> Finished dependency resolution
Error:
 Problem 1: conflicting requests
  - nothing provides /usr/bin/env needed by dnf-4.2.2-r0.aarch64
 Problem 2: conflicting requests
  - nothing provides /usr/bin/env needed by rpm-1:4.14.2.1-r0.aarch64
 Problem 3: package matchbox-terminal-0.2-r0.aarch64 requires libgtk-3.so.0()(64bit), but none of the providers can be installed
  - package matchbox-terminal-0.2-r0.aarch64 requires libgtk-3.0 >= 3.24.14, but none of the providers can be installed
  - package packagegroup-core-x11-base-1.0-r1.noarch requires matchbox-terminal, but none of the providers can be installed
  - package libgtk-3.0-3.24.14-r0.aarch64 requires glib-2.0-utils, but none of the providers can be installed
  - conflicting requests
  - nothing provides /usr/bin/env needed by libglib-2.0-utils-1:2.62.6-r0.aarch64
 Problem 4: package packagegroup-core-x11-sato-base-1.0-r33.jetson_xavier_nx_devkit_tx2_nx requires settings-daemon, but none of the providers can be installed
  - package packagegroup-core-x11-sato-1.0-r33.jetson_xavier_nx_devkit_tx2_nx requires packagegroup-core-x11-sato-base, but none of the providers can be installed
  - package settings-daemon-0.0.2-r0.aarch64 requires libgconf-2.so.4()(64bit), but none of the providers can be installed
  - package settings-daemon-0.0.2-r0.aarch64 requires gconf >= 3.2.6, but none of the providers can be installed
  - conflicting requests
  - nothing provides /usr/bin/env needed by gconf-3.2.6-r0.aarch64
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

ERROR: Logfile of failure stored in: /home/synxbuild/yocto-tegra/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-poky-linux/core-image-sato-dev/1.0-r0/temp/log.do_rootfs.3724324
ERROR: Task (/home/synxbuild/yocto-tegra/poky-dunfell/meta/recipes-sato/images/core-image-sato-dev.bb:do_rootfs) failed with exit code '1'
```

**Reply by madisongh** [reply]
> Do you know if the yocto-tegra build includes the proper dependencies necessary for building a simple C++ program?

There's nothing special about meta-tegra in this regard; it just provides recipes to support the Jetson boards.  However, as noted in the [README](https://github.com/OE4T/meta-tegra/tree/dunfell#please-note) on the dunfell branch, if you plan on building any CUDA code, you must use an older version of the gcc/g++ cross-compiler than the default provided in OE-Core.

But the error messages you included have nothing to do with compiling C++ code; they are errors generated when trying to construct the rootfs for, apparently, `core-image-sato-dev`.  If you can be more specific about exactly what revision/branch each of your layers is at (the initial output from `bitbake` shows this) and exactly what your build command was, that might help in trying to track down why you're seeing these errors.

We also provide a [demo distro](https://github.com/OE4T/tegra-demo-distro/tree/dunfell) to make it easier to get started; you might find it helpful to try a build using that for comparing against your current setup to debug build issues.

**Reply by madisongh** [reply]
And just a note that I tried a `core-image-sato-dev` build in a basic poky setup that went OK:
```
$ bitbake core-image-sato-dev
WARNING: Host distribution "ubuntu-21.10" has not been validated with this version of the build system; you may possibly experience unexpected failures. It is recommended that you use a tested distribution.
Parsing recipes: 100% |##############################################################################################| Time: 0:00:02
Parsing of 907 .bb files complete (0 cached, 907 parsed). 1509 targets, 60 skipped, 0 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.46.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-poky-linux"
MACHINE              = "jetson-xavier-nx-devkit-tx2-nx"
DISTRO               = "poky"
DISTRO_VERSION       = "3.1.13"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            
meta-yocto-bsp       = "dunfell:3432003435a2b86d72a1438f497032959c99b672"
meta-tegra           = "dunfell:9233789562c3be0bfd201d6f1a9fb927fd2d257b"

Initialising tasks: 100% |###########################################################################################| Time: 0:00:03
Sstate summary: Wanted 2529 Found 66 Missed 2463 Current 543 (2% match, 19% complete)
NOTE: Executing Tasks
NOTE: Tasks Summary: Attempted 7485 tasks of which 2011 didn't need to be rerun and all succeeded.

Summary: There was 1 WARNING message shown.
```
The warning is just because my build host is running a newer version of Ubuntu than the Yocto Project used.

**Comment by tmdarwen** [comment]
Thanks @madisongh - I wasn't aware of the demo distro.  I've built it and managed to flash it to the Jetson but when the Jetson boots a splash screen looking [like this](https://avatars.githubusercontent.com/u/69109450?s=200&v=4) appears and never goes away, no mouse control or anything.

I think I just realized my problem. To build, I ran `bitbake core-image-sato-dev` and should've run `bitbake demo-image-sato`

**Reply by ichergui** [reply]
@tmdarwen 

You can try `bitbake demo-image-sato` 
Please share the serial uart logs

**Comment by tmdarwen** [comment, marked as answer]
Thanks guys, it turned out to be a typo I had in my local.conf to add my example C++ layer.  I had added `IMAGE_INSTALL = "example"` when it should have been `IMAGE_INSTALL_append = " example"`.

My "example" C++ application is now being built and added to the demo image. 

When starting up the Jetson with the demo image the example executable is in /usr/bin and displays the expected "hello world" output.

Thank you for the help...  I'm getting there....
